### ISSUE TITLE
Libtorch cpp interface tracing error

### ISSUE THREAD

**Issue by zhiqwang** [facebookresearch/detr#238]
## Instructions To Reproduce the 🐛 Bug:

1. what changes you made (`git diff`) or what code you wrote

- I use the libtorch cpp interface to load the exported torchscript model. Concrete code as below:

```cpp
#include <ATen/ATen.h>
#include <torch/script.h>
#include <torch/torch.h>


int main() {
  torch::DeviceType device_type;
  device_type = torch::kCPU;

  torch::jit::script::Module module;
  try {
    std::cout << "Loading model" << std::endl;
    // Deserialize the ScriptModule from a file using torch::jit::load().
    module = torch::jit::load("detr_resnet50.pt");
    std::cout << "Model loaded" << std::endl;
  } catch (const torch::Error& e) {
    std::cout << "error loading the model" << std::endl;
    return -1;
  } catch (const std::exception& e) {
    std::cout << "Other error: " << e.what() << std::endl;
    return -1;
  }

  // TorchScript models require a List[IValue] as input
  std::vector<torch::jit::IValue> inputs;

  // Demonet accepts a List[Tensor] as main input
  std::vector<torch::Tensor> images;
  images.push_back(torch::rand({3, 200, 200}));
  images.push_back(torch::rand({3, 256, 275}));

  inputs.push_back(images);
  auto output = module.forward(inputs);

  std::cout << "ok" << std::endl;
  std::cout << "output" << output << std::endl;
  return 0;
}
```

- And my CMakeLists.txt

```cmake
cmake_minimum_required(VERSION 3.1 FATAL_ERROR)
project(test_tracing)

find_package(Torch REQUIRED)

# This due to some headers importing Python.h
find_package(Python3 COMPONENTS Development)

add_executable(${CMAKE_PROJECT_NAME} test_tracing.cpp)
target_compile_features(test_tracing PUBLIC cxx_range_for)

target_link_libraries(
  ${CMAKE_PROJECT_NAME}
  ${TORCH_LIBRARIES}
  Python3::Python
)

# set C++14 to compile
set_property(TARGET test_tracing PROPERTY CXX_STANDARD 14)

```

2. what exact command you run:

```shell
mkdir build && cd build
cmake .. -DTorch_DIR=$TORCH_PATH/share/cmake/Torch
make
./test_tracing
```

3. what you observed (including __full logs__):

```shell
Loading model
Model loaded
terminate called after throwing an instance of 'c10::Error'
  what():  forward() Expected a value of type '__torch__.util.misc.NestedTensor' for argument 'samples' but instead found type 'List[Tensor]'.
Position: 1
Declaration: forward(__torch__.models.detr.DETR self, __torch__.util.misc.NestedTensor samples) -> (Dict(str, Tensor))
Exception raised from checkArg at /pytorch/aten/src/ATen/core/function_schema_inl.h:193 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x42 (0x7f1bb76cb572 in /usr/local/lib/python3.6/dist-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0xb6d218 (0x7f1ba7b94218 in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch_cpu.so)
frame #2: <unknown function> + 0xb72017 (0x7f1ba7b99017 in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch_cpu.so)
frame #3: torch::jit::GraphFunction::operator()(std::vector<c10::IValue, std::allocator<c10::IValue> >, std::unordered_map<std::string, c10::IValue, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, c10::IValue> > > const&) + 0x2d (0x7f1ba9c23a0d in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch_cpu.so)
frame #4: torch::jit::Method::operator()(std::vector<c10::IValue, std::allocator<c10::IValue> >, std::unordered_map<std::string, c10::IValue, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, c10::IValue> > > const&) + 0x146 (0x7f1ba9c321a6 in /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch_cpu.so)
frame #5: torch::jit::Module::forward(std::vector<c10::IValue, std::allocator<c10::IValue> >) + 0xf9 (0x5618878b0ca5 in ./test_tracing)
frame #6: main + 0x282 (0x5618878ac14d in ./test_tracing)
frame #7: __libc_start_main + 0xe7 (0x7f1b6a4abb97 in /lib/x86_64-linux-gnu/libc.so.6)
frame #8: _start + 0x2a (0x5618878ab60a in ./test_tracing)
```

4. please simplify the steps as much as possible so they do not require additional resources to
	 run, such as a private dataset.

## Expected behavior:

The libtorch cpp interface can do inference :)

## Environment:

Provide your environment information using the following command:
```
Collecting environment information...
PyTorch version: 1.7.0.dev20200912
Is debug build: False
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.3 LTS (x86_64)
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
Clang version: 11.0.0 (git@gitee.com:JoelYang/llvm-project.git b3fb40b3a3c1fb7ac094eda50762624baad37552)
CMake version: version 3.18.0

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration:
GPU 0: Tesla P100-SXM2-16GB

Nvidia driver version: 384.81
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.8.0.3
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.3
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.3
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.3
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.3
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.3
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.3

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.7.0.dev20200912
[pip3] torchvision==0.8.0.dev20200914
[conda] Could not collect
```

**Comment by zhiqwang** [comment]
Hi,

It seems the method

https://github.com/facebookresearch/detr/blob/5e66b4cd15b2b182da347103dd16578d28b49d69/util/misc.py#L286-L289

and the input of `DETR` model

https://github.com/facebookresearch/detr/blob/5e66b4cd15b2b182da347103dd16578d28b49d69/models/detr.py#L59-L60

will cause the the tracing failed.

My initial idea is specific defined the type of the inputs, just use type `List[torch.Tensor]` to avoid this error. And I did an experiment in my fork [here](https://github.com/zhiqwang/detr/commit/7739b31a22ecd8c9ec828e136e209597a205454c). This modification can handle the libtorch tracing in the cpp interface.

Is there more suggestion to fix this problem? If my modification works, I would like to submit a PR :)

**Comment by fmassa** [comment]
Hi,

In order to use the code in C++ without changes, you would need to wrap your input in a NestedTensor, which is what torchscript expects.

Here are two options:
# 1- wrap your DETR model in a dummy model that takes a `List[Tensor]` as follows
```python
class WrappedDETR(nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model

    def forward(self, inputs : List[Tensor]):
        return self.model(inputs)
```
This should be the easiest way to do it I think.

# 2 - export `nested_tensor_from_tensor_list` as a separate torchscript object, and use it to convert your list of tensors into a NestedTensor. This adds a bit more complexity on the C++ side of things.

I'm not sure if there is a simpler way to construct the NestedTensor object from C++, but I believe those two options should work.

For more context on the existence of NestedTensor, I suggest you check https://github.com/facebookresearch/detr/issues/116. We would like to keep the `NestedTensor` abstractions in the constructor for now, to avoid potentially breaking changes on user code.

I believe I've answered your question, and as such I'm closing this issue, but let us know if you have further questions.

**Comment by zhiqwang** [comment]
Hi @fmassa , thanks for you quickly reply.

I try your first suggestions, and there are subsequent error followed up this modification.

My detailed torchscript exporting scripts as below:

```python
# file: trace_model.py

from hubconf import detr_resnet50


model = detr_resnet50(pretrained=False)
wrapped_model = WrappedDETR(model)
wrapped_model.eval()

traced_model = torch.jit.script(wrapped_model)
traced_model.save("detr_resnet50.pt")
```

These caused following error happened:

```shell
forward(__torch__.models.detr.DETR self, __torch__.util.misc.NestedTensor samples) -> (Dict(str, Tensor)):
Expected a value of type '__torch__.util.misc.NestedTensor' for argument 'samples' but instead found type 'List[Tensor]'.
Empty lists default to List[Tensor]. Add a variable annotation to the assignment to create an empty list of another type (torch.jit.annotate(List[T, []]) where T is the type of elements in the list for Python 2)
:
  File "trace_model.py", line 19
    def forward(self, inputs: List[Tensor]):
        return self.model(inputs)
               ~~~~~~~~~~ <--- HERE

root@4694ab6195af:/mnt/coding/detr# python trace_model.py
Traceback (most recent call last):
  File "trace_model.py", line 26, in <module>
    traced_model = torch.jit.script(wrapped_model)
  File "/usr/local/lib/python3.6/dist-packages/torch/jit/_script.py", line 893, in script
    obj, torch.jit._recursive.infer_methods_to_compile
  File "/usr/local/lib/python3.6/dist-packages/torch/jit/_recursive.py", line 338, in create_script_module
    return create_script_module_impl(nn_module, concrete_type, stubs_fn)
  File "/usr/local/lib/python3.6/dist-packages/torch/jit/_recursive.py", line 397, in create_script_module_impl
    create_methods_from_stubs(concrete_type, stubs)
  File "/usr/local/lib/python3.6/dist-packages/torch/jit/_recursive.py", line 291, in create_methods_from_stubs
    concrete_type._create_methods(defs, rcbs, defaults)
RuntimeError:

forward(__torch__.models.detr.DETR self, __torch__.util.misc.NestedTensor samples) -> (Dict(str, Tensor)):
Expected a value of type '__torch__.util.misc.NestedTensor' for argument 'samples' but instead found type 'List[Tensor]'.
Empty lists default to List[Tensor]. Add a variable annotation to the assignment to create an empty list of another type (torch.jit.annotate(List[T, []]) where T is the type of elements in the list for Python 2)
:
  File "trace_model.py", line 19
    def forward(self, inputs: List[Tensor]):
        return self.model(inputs)
               ~~~~~~~~~~ <--- HERE
```

It seems that the trace here can only be recognized as `NestedTensor` type. So I modify your supplied code snippets to:

```python
from util.misc import nested_tensor_from_tensor_list


class WrappedDETR(nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model

    def forward(self, inputs: List[Tensor]):
        sample = nested_tensor_from_tensor_list()
        return self.model(sample)
```

This changes works, but also cause the following error,

```shell
_onnx_nested_tensor_from_tensor_list(Tensor tensor_list) -> (Tensor 0):
Expected a value of type 'Tensor' for argument 'tensor_list' but instead found type 'List[Tensor]'.
Empty lists default to List[Tensor]. Add a variable annotation to the assignment to create an empty list of another type (torch.jit.annotate(List[T, []]) where T is the type of elements in the list for Python 2)
:
  File "/mnt/coding/detr/util/misc.py", line 289
            # nested_tensor_from_tensor_list() does not export well to ONNX
            # call _onnx_nested_tensor_from_tensor_list() instead
            return _onnx_nested_tensor_from_tensor_list(tensor_list)
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE

        # TODO make it support different-sized images
'nested_tensor_from_tensor_list' is being compiled since it was called from 'WrappedDETR.forward'
  File "trace_model.py", line 18
    def forward(self, inputs: List[Tensor]):
        sample = nested_tensor_from_tensor_list()
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
        return self.model(sample)
```

This happed from following location

https://github.com/facebookresearch/detr/blob/5e66b4cd15b2b182da347103dd16578d28b49d69/util/misc.py#L286-L289

Which can be solved by following modification to make the output type in this control flow unified.(the detailed code is in [here](https://github.com/zhiqwang/detr/blob/0572c9729294406918df5e241adc03832e6b60bd/util/misc.py#L283-L342)). And even if use other methods in my own branches as my try, these bugs would happened too. What's your opinions here?

```python
def nested_tensor_from_tensor_list(tensor_list: List[Tensor]):
    # TODO make this more general
    if tensor_list[0].ndim == 3:
        if torchvision._is_tracing():
            # _nested_tensor_from_tensor_list() does not export well to ONNX
            # call _onnx_nested_tensor_from_tensor_list() instead
            tensor, mask = _onnx_nested_tensor_from_tensor_list(tensor_list)
        else:
            tensor, mask = _nested_tensor_from_tensor_list(tensor_list)
    else:
        raise ValueError('not supported')
    return NestedTensor(tensor, mask)
```

**Comment by fmassa** [comment]
@zhiqwang is the only difference the fact that you wrap only once in a NestedTensor?

**Comment by zhiqwang** [comment]
Hi @fmassa , I only add a line of your supplied `WrappedDETR` class to suppress a RuntimeError:

```python
class WrappedDETR(nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model

    def forward(self, inputs: List[Tensor]):
        sample = nested_tensor_from_tensor_list(inputs)
        return self.model(sample)
```

But it seems that there is a hidden bug in function `nested_tensor_from_tensor_list`. The control flow as below will cause the typing incorrect. The returned type of two branches are not consistent. (Line 289 and L305)

https://github.com/facebookresearch/detr/blob/5e66b4cd15b2b182da347103dd16578d28b49d69/util/misc.py#L283-L305

So I just let the function `_onnx_nested_tensor_from_tensor_list` return `tensor` and `mask`, and sent them into `NestedTensor(tensor, mask)`. Something like below, the full code is in [here](https://github.com/zhiqwang/detr/blob/9911589d1ac9e08b532c11d76cb946a0b1cea742/util/misc.py#L283-L342).

```python
def nested_tensor_from_tensor_list(tensor_list: List[Tensor]):
    # TODO make this more general
    if tensor_list[0].ndim == 3:
        if torchvision._is_tracing():
            # _nested_tensor_from_tensor_list() does not export well to ONNX
            # call _onnx_nested_tensor_from_tensor_list() instead
            tensor, mask = _onnx_nested_tensor_from_tensor_list(tensor_list)
        else:
            tensor, mask = _nested_tensor_from_tensor_list(tensor_list)
    else:
        raise ValueError('not supported')
    return NestedTensor(tensor, mask)
```

After this modification, all thing works.

**Comment by fmassa** [comment]
Thanks for the detailed explanation, that was indeed my question.

I now understand the problem: in 
https://github.com/facebookresearch/detr/blob/5e66b4cd15b2b182da347103dd16578d28b49d69/util/misc.py#L311
We need to set the type annotation to be `List[Tensor]`, otherwise it defaults to `Tensor`.

Can you try this out, and if it works send a PR fixing it?

**Comment by zhiqwang** [comment]
@fmassa It's my pleasure, But just add the type annotation tp be `List[Tensor]` in Line 311 will also cause this RuntimeError

```shell
RuntimeError:
Previous return statement returned a value of type Tensor but this return statement returns a value of type __torch__.util.misc.NestedTensor:
  File "/mnt/coding/detr/util/misc.py", line 305
    else:
        raise ValueError('not supported')
    return NestedTensor(tensor, mask)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
'nested_tensor_from_tensor_list' is being compiled since it was called from 'WrappedDETR.forward'
  File "trace_model.py", line 18
    def forward(self, inputs: List[Tensor]):
        sample = nested_tensor_from_tensor_list(inputs)
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
        return self.model(sample)
```

**Comment by fmassa** [comment]
Weird, this seems like a bug with the type propagation in the function.

Do you have a code-snippet that reproduces this issue? Also, can you already at least send a PR fixing the type annotation of the ONNX counterpart of the function? This way we can see how CI behaves with it.

**Comment by zhiqwang** [comment]
Hi fmassa, Okay, I'll submit a PR.

**Comment by zhiqwang** [comment]
Hi @fmassa I write a unittest to let CI reproduces this issue, I open it in another PR?

**Comment by fmassa** [comment]
@zhiqwang thanks for the PR! I have one hypothesis why the tests fail that I commented in the PR, let's move the discussion there.
